### PR TITLE
Open Dynamo Template as new workspace

### PR DESCRIPTION
**Pull request by QilongTang**
### Purpose

Introduce open by template function in Dynamo.

ToDo:

- [ ] Opening a template then save should not save to template path (This may require another PR to add a new property to current workspace)
- [x] Trying to save to template path should be blocked with message dialog
- [ ] Short cut key combo Ctrl + T does not work yet

image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section**

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: test/DynamoCoreWpfTests/PreferencesViewModelTests.cs [hunk 1]
```diff
@@ -157,10 +157,12 @@ public void PathManagerWithDifferentHostTest()
             string dynamoRevitUserDataDirectory = "C:\\Users\\user\\AppData\\Roaming\\Dynamo\\Dynamo Revit\\3.1";
             string dynamoRevitCommonDataDirectory = "C:\\ProgramData\\Autodesk\\RVT 2024\\Dynamo\\3.1";
             string dynamoRevitSamplesPath = "C:\\ProgramData\\Autodesk\\RVT 2024\\Dynamo\\samples\\en-US";
+            string dynamoRevitTemplatesPath = "C:\\ProgramData\\Autodesk\\RVT 2024\\Dynamo\\templates\\en-US";
```
Comment (RobertGlobant20) [inline test/DynamoCoreWpfTests/PreferencesViewModelTests.cs:160]: You could refactor this code with `Path.Combine` to use the common directory "C:\\ProgramData\\Autodesk\\RVT 2024\\Dynamo"
Something like:
string dynamoRevitCommonDataDirectory = Path.Combine(RVTForDynamoDir,"3.1") ;

**Review by RobertGlobant20 — APPROVED** [review]
LGTM with some comments

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -1957,6 +1957,30 @@ public void OpenFileFromPath(string filePath, bool forceManualExecutionMode = fa
...
+        /// <param name="filePath">Path to file</param>
+        /// <param name="forceManualExecutionMode">Set this to true to discard
+        /// execution mode specified in the file and set manual mode</param>
+        public void OpenTemplateFromPath(string filePath, bool forceManualExecutionMode = false)
+        {
+
+            if (DynamoUtilities.PathHelper.isValidJson(filePath, out string fileContents, out Exception ex))
+            {
+                OpenJsonFileFromPath(fileContents, filePath, forceManualExecutionMode, true);
+                return;
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:1972]: Will remove

File: src/DynamoCore/Models/DynamoModel.cs [hunk 3]
```diff
@@ -2040,7 +2065,7 @@ private bool OpenJsonFileFromPath(string fileContents, string filePath, bool for
                     if (true) //MigrationManager.ProcessWorkspace(dynamoPreferences.Version, xmlDoc, IsTestMode, NodeFactory))
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:2065]: Good question! I dont.. looks like another migration check was left behind

File: src/DynamoCore/Models/DynamoModel.cs [hunk 4]
```diff
@@ -2295,7 +2321,7 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
                 CustomNodeManager,
                 this.LinterManager);
 
-            workspace.FileName = string.IsNullOrEmpty(filePath) ? "" : filePath;
+            workspace.FileName = string.IsNullOrEmpty(filePath) || isTemplate? "" : filePath;
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply src/DynamoCore/Models/DynamoModel.cs:2324]: Can do

File: test/DynamoCoreWpfTests/PreferencesViewModelTests.cs (hunk 1 above)
Comment (QilongTang) [reply test/DynamoCoreWpfTests/PreferencesViewModelTests.cs:160]: I can update the test

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 5]
```diff
@@ -1964,6 +1973,62 @@ private void ShowOpenDialogAndOpenResult(object parameter)
 
         private bool CanShowOpenDialogAndOpenResultCommand(object parameter) => CanRunGraph;
 
+        /// <summary>
+        /// Present the open dialog and open the template that is selected.
+        /// </summary>
+        /// <param name="parameter"></param>
+        private void ShowOpenTemplateDialog(object parameter)
```
Comment (zeusongit) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:1980]: Any way we can combine this with `ShowOpenDialogAndOpenResult `?

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:1980]: yeah I thought about this but we need to call `Open` function below with different tuple and use `TemplatesDirectory` as start directory, currently there is nothing we can use to differentiate two workflows other than use different command for each menu item. I can try think about this more

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 6]
```diff
@@ -334,9 +337,18 @@
...
+                                  Name="openButton">
+                            <MenuItem Header="{x:Static p:Resources.DynamoViewFileMenuOpenFile}"
+                                      Name="openFileMenuItem"
+                                      Command="{Binding ShowOpenDialogAndOpenResultCommand}"
+                                      InputGestureText="Ctrl + O">
+                            </MenuItem>
+                            <MenuItem Header="{x:Static p:Resources.DynamoViewFileMenuOpenTemplate}"
+                                      Name="openTemplateMenuItem"
+                                      Command="{Binding ShowOpenTemplateDialogCommand}"
+                                      CommandParameter="Template"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml:349]: @zeusongit I decided to use command param so we can distinguish if Dynamo is opening a template so we can reuse more file open code

**Comment by QilongTang** [comment]
Merging this PR in to create the base of template function.

**Pull request merged**